muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,682 @@
+"""
+Autoencoder in scikit-learn style with TensorFlow
+"""
+import logging
+import re
+import warnings
+
+import numpy as np
+import tensorflow as tf
+import scipy.sparse as sp
+
+from sklearn.base import TransformerMixin, BaseEstimator
+from sklearn.utils import check_array, check_random_state
+from sklearn.utils.validation import NotFittedError
+
+from ..core import TFPicklingBase, affine
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Autoencoder(TFPicklingBase, TransformerMixin, BaseEstimator):
+    """
+    An autoencoder using TensorFlow.
+
+    Parameters
+    ----------
+    hidden_units : tuple or list, optional
+        A list of integers indicating the number of hidden layers and their
+        sizes. The list is reflected about the last element to form the
+        autoencoder. For example, if hidden_units is (256, 32, 16) then
+        the autoencoder has layers
+
+            [input_layer, 256, 32, 16, 32, 256, input_layer]
+
+    batch_size : int, optional
+        The batch size for learning and prediction. If there are fewer
+        examples than the batch size during fitting, then the the number of
+        examples will be used instead.
+    n_epochs : int, optional
+        The number of epochs (iterations through the training data) when
+        fitting.
+    dropout : float or None, optional
+        The dropout probability. If None, then dropout will not be used.
+        Note that dropout is not applied to the input layer.
+    activation : (callable, callable), optional
+        The activation functions for
+
+            (the bulk of the autoencoder and encoding layer, output layer).
+
+        See `tensorflow.nn` for various options.
+    random_state: int, RandomState instance or None, optional
+        If int, the random number generator seed. If RandomState instance,
+        the random number generator itself. If None, then `np.random` will be
+        used.
+    learning_rate : float, optional
+        Learning rate for Adam.
+    metric : string, optional
+        Default metric for the autoencoder. Options are
+
+            'mse' - mean square error
+            'cross-entropy' - cross-entropy
+
+        Note that the 'cross-entropy' metric forces the activation of the
+        output layer to be tensorflow.nn.sigmoid. Any dimensions indicated by
+        the options `categorical_begin`, `categorical_size` or
+        `discrete_indices` use a cross-entropy metric with
+        `tensorflow.nn.softmax` or `tensorflow.nn.sigmoid` output layer
+        activations.
+    discrete_indices : array-like, 1d, optional
+        Array of indices for which `tf.nn.sigmoid` will be used for the
+        output layer activation and the cross-entropy will be used as the
+        metric for the autoencoder.
+    categorical_begin : array-like, shape (n_categorial), optional
```

Comment:
Sure I can make this change. I like the current structure since it is more explicit and you don't have to remember the order (i.e. is it `(begin, size)` or `(size, begin)`?)